File: Go/cryptography/Vernam_Cipher/Vernam_Cipher.go
/* This code is an implementation of Vernam Cipher in Go. In
Vernam cipher, also known as the One-Time Pad, the length of
the plaintext, ciphertext, and key is the same. Basically,
here the ith letter of the cipher text is fomred by shifting
the ith letter of the plaintext with the ith letter of key.
*/

package main

import (
	"fmt"
	"unicode"
	"strings"
	"os"
)

// This is the encryption function
func encryption(plaintext, key string) string {

	var encrypted string

	// Shifting the plaintext to generate the ciphertext
	for i := 0; i < len(plaintext); i++ {
		var temp int = (int(plaintext[i]) - 97 + int(key[i]) - 97) % 26
		temp = temp + 97
		encrypted += string(temp)
	}
	// Returns the ciphertext
	return encrypted
}

func format(plaintext, ciphertext string) string {

	var enc string

	/* Check wherever the plaintext is capital and
	then capitalize that character in the ciphertext*/
	for i := 0; i < len(plaintext); i++ {
		if(unicode.IsUpper(rune(plaintext[i]))) {
			enc += strings.ToUpper(string(ciphertext[i]))
		} else {
			enc += string(ciphertext[i])
		}
	}
	// Return the final formatted ciphertext
	return enc
}

func main() {

	// Taking plaintext as input from the user
	fmt.Print("Enter a message to encrypt: ")
	var plaintext string 
	fmt.Scan(&plaintext)

	// Taking key as input from the user
	fmt.Print("Enter a key of same length: ")
	var key string
	fmt.Scan(&key)

	/* If length of key is not equal to length
 	   of plaintext then exit the program.*/
	if(len(key) != len(plaintext)) {
		fmt.Print("\nInvalid key size.\n")
		os.Exit(0)
	}

	// Generating the ciphertext
	var ciphertext string = encryption(strings.ToLower(plaintext), strings.ToLower(key))

	// Formatting the ciphertext
	var encrypted string = format(plaintext, ciphertext)
	fmt.Print("\nEncrypted plaintext is    : ", encrypted, "\n")
}

/* Sample I/O:

a)
Enter a message to encrypt: ThisIsAMsg
Enter a key of same length: ThisisaKey

Encrypted plaintext is    : MoqkQkAWwe

b)
Enter a message to encrypt: PlaintextMessage
Enter a key of same length: Key

Invalid key size.
*/

File: Java/cryptography/vernamCipher.java
/*In Vernam cipher, also known as the One-Time Pad, the length
  of the plaintext, ciphertext, and key is the same. Below is
  an implementation of Vernam cipher in Java. Here the ith letter
  of the cipher text is fomred by shifting the numerical value of
  ith letter of the plaintext by the numerical value ofith letter
  of key. The conversion here is a-0, b-1, c-2, ...., z-25.*/

import java.util.*;
import java.lang.*;

public class vernamCipher {

	/* This method takes a string as input and converts
	   each character in the string to its respective
	   numerical value and returns it.*/
	static int[] asciiToDec(String str) {

		int[] decimal = new int[str.length()];

		for(int i = 0; i < str.length(); i++) {

			/* We first convert the character to its
			 decimal and then shift it by 97 as, a-97,
			 b-98, and so on. So to change the indexing
			 and start from 0, we subtract the value by 97.*/
			decimal[i] = str.charAt(i);
			decimal[i] = decimal[i] - 97;
		}

		return decimal;
	}

	/* This method is the encryption method. It takes 
	   plaintext and key as the input.*/
	static String encrypt(String plaintext, String key) {

		String ciphertext = new String("");

		/* Converting both plaintext and key to their
		   respective numerical values.*/
		int[] plaintextArray = asciiToDec(plaintext);
		int[] keyArray = asciiToDec(key);

		int[] ciphertextArray = new int[keyArray.length];

		/* Forming the ciphertext by shifting each value of
		plaintext by the value present at same index of the key*/
		for(int i = 0; i < keyArray.length; i++) {
			ciphertextArray[i] = (plaintextArray[i] + keyArray[i]) % 26;

			//Shifting the values by 97, to get proper ascii values.
			ciphertextArray[i] += 97;
		}

		/* Converting the decimal values to respective ascii
		values and forming the ciphertext string.*/
		for(int i = 0; i < ciphertextArray.length; i++) {
			ciphertext += (char)ciphertextArray[i];
		}

		return ciphertext;
	}

	/* This method formats the ciphertext according to the plaintext.
	From the encrypt function above, we send plaintext in lowercase.
	So, the ciphertext we get is also in all lowercase. Here, we format
	the ciphertext so that wherever there is capital letter in the plaintext,
	same follows in the ciphertext.*/
	static String format(String plaintext, String ciphertext) {

		String finalencrypted = new String("");

		/* Check wherever the plaintext is capital and
		then capitalize that character in the ciphertext*/
		for(int i = 0; i < plaintext.length(); i++) {
			if(Character.isUpperCase(plaintext.charAt(i))) {
				finalencrypted += Character.toUpperCase(ciphertext.charAt(i));
			}
			else {
				finalencrypted += ciphertext.charAt(i);
			}
		}
		return finalencrypted;
	}

	public static void main(String[] args) {

		Scanner scan = new Scanner(System.in);

		// Take plaintext and key as input from the user.
		System.out.print("Enter a plaintext to encrypt: ");
		String plaintext = scan.nextLine();

		System.out.print("Enter a key of same size    : ");
		String key = scan.nextLine();

		/*If the length of key is not equal to the
		length of the plaintext, then terminate.*/
		if(plaintext.length() != key.length()) {
			System.out.println("Invalid input.");
			System.exit(0);
		}
		scan.close();

		// Call the methods and print the encrypted message.
		String encryptedplaintext = encrypt(plaintext.toLowerCase(), key.toLowerCase());
		String encrypted = format(plaintext, encryptedplaintext);

		System.out.println("Encrypted plaintext is      : " + encrypted);
	}
}

/* Sample I/O:
a)	Enter a plaintext to encrypt: ThisIsAMessage
	Enter a key of same size    : MercuryisHottt
	Encrypted plaintext is      : FlzuCjYUwzgtzx

b)	Enter a plaintext to encrypt: MessageTooBIg
	Enter a key of same size    : KeynotMatch
	Invalid input.
*/

